numpy/core/src/multiarray/mapping.c

```diff
@@ -235,29 +206,79 @@ prepare_index(PyArrayObject *self, PyObject *index,
 
         if (make_tuple) {
             /* We want to interpret it as a tuple, so make it one */
-            index = PySequence_Tuple(index);
-            if (index == NULL) {
-                return -1;
+            index_as_tuple = PySequence_Tuple(index);
+            if(index_as_tuple == NULL) {
+                return NULL;
             }
         }
     }
 
-    /* If the index is not a tuple, handle it the same as (index,) */
-    if (!PyTuple_CheckExact(index)) {
-        obj = index;
-        index_ndim = 1;
+    /* If the index is not a tuple, conver it into (index,) */
+    if (!make_tuple && !PyTuple_CheckExact(index)) {
+        index_as_tuple = Py_BuildValue("(O)", index);
     }
+    /* Otherwise, check if the tuple is too long */
     else {
-        n = PyTuple_GET_SIZE(index);
+        n = PyTuple_GET_SIZE(index_as_tuple);
         if (n > NPY_MAXDIMS * 2) {
             PyErr_SetString(PyExc_IndexError,
                             "too many indices for array");
-            goto fail;
+            if(make_tuple) {
+                Py_DECREF(index_as_tuple);
+            }
+            return NULL;
         }
-        index_ndim = (int)n;
-        obj = NULL;
     }
 
+    return index_as_tuple;
+}
+
+/**
+ * Prepare an npy_index_object from the python slicing object.
+ *
+ * This function handles all index preparations with the exception
+ * of field access. It fills the array of index_info structs correctly.
+ * It already handles the boolean array special case for fancy indexing,
+ * i.e. if the index type is boolean, it is exactly one matching boolean
+ * array. If the index type is fancy, the boolean array is already
+ * converted to integer arrays. There is (as before) no checking of the
+ * boolean dimension.
+ *
+ * Checks everything but the bounds.
+ *
+ * @param the array being indexed
+ * @param the index object
+ * @param index info struct being filled (size of NPY_MAXDIMS * 2 + 1)
+ * @param number of indices found
+ * @param dimension of the indexing result
+ * @param dimension of the fancy/advanced indices part
+ * @param whether to allow the boolean special case
+ *
+ * @returns the index_type or -1 on failure and fills the number of indices.
+ */
+NPY_NO_EXPORT int
+prepare_index(PyArrayObject *self, PyObject *index_raw,
```

Comment:
This became `index_raw` because I did not fancy my odds of correctly renaming all the existing variables. I'll do that in a separate commit